urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/httpd/resource/RangedResource.java

```diff
@@ -0,0 +1,203 @@
+import ikrs.io.ReadLimitInputStream;
+import ikrs.io.fileio.htaccess.HypertextAccessFile;
+
+/**
+ * The RangedResource class is a wrapper class that allows to 'cut off' some
+ * bytes at the beginning and/or at the end of an existing resource.
+ *
+ * The amount of bytes to be dropped is specified by a ContentRange instance
+ * passed to the constructor.
+ *
+ * Actually the RangedResource uses a custom getInputStream() replacement
+ * with a nested ReadLimit- (to drop trailing bytes) and BytePosition-InputStream 
+ * (to drop leading bytes).
+ *
+ *
+ * @author Ikaros Kappler
+ * @date 2013-02-27
+ * @version 1.0.0
+ **/
+
+
+public class RangedResource
+    extends ResourceDelegation {
+
+    private ContentRange contentRange;
+
+    private ReadLimitInputStream readLimitInputStream;
+
+    public RangedResource( Resource resource,
+			   ContentRange range,
+			   
+			   HTTPHandler handler,
+			   CustomLogger logger ) 
+	throws NullPointerException,
+	       IndexOutOfBoundsException {
+
+	super( resource,
+	       
+	       handler,
+	       logger
+	       );
+
+	if( range == null )
+	    throw new NullPointerException( "Cannot create a RangedResource with null-ContentRange." );
+
+	this.contentRange            = range;
+	this.readLimitInputStream    = null;         // will be initialized on open(...)
+    }
+
+
+    //--- BEGIN --------------------- Override ResourceDelegation -------------------------
+    /**
+     * This method opens the underlying resource. Don't forget to close.
+     *
+     * @param readOnly if set to true, the resource will be opned in read-only mode.
+     *
+     * @throws ReadOnlyException If the underlying resource is read-only in general.
+     * @throws IOException If any other IO error occurs.
+     * @see isReadOnly()
+     **/
+    public void open( boolean readOnly )
+	throws IOException {
+
+	// Open the core resource first!
+	// (otherwise we will not be able to access the input stream)
+	super.open( readOnly );
+      
+	// No need to store the dat of last modification:
+	//  both resources share the same meta data instance.
+    }
+
+    /**
+     * Get the input stream from this resource.
+     *
+     * @throws IOException If any IO error occurs.
+     **/
+    public InputStream getInputStream()
+	throws IOException {
+
+	if( !this.isOpen() )
+	    throw new IOException( "Cannot retrieve resource's input stream. Resouce was not yet opened." );
+
+	if( this.readLimitInputStream == null ) {
+
+	    if( this.contentRange.getLastBytePosition() >= this.getCoreResource().getLength() )
+		throw new IOException( "Failed to initialize the RangedResource's input stream; the passed Content-Range " + this.contentRange.toString() + " is out of bounds!" );
+
+	    // Init the stream(s)!
+	    InputStream tmpIn = this.getCoreResource().getInputStream();
+	    if( this.contentRange.getFirstBytePosition() > 0 ) {
+		
+		// Skip n bytes at the beginning
+		tmpIn.skip( this.contentRange.getFirstBytePosition() );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Check the number of bytes that were skipped when you called `skip()`. Handle cases when fewer than the expected number of bytes were skipped. See the following diffs on GitHub that make this code change: [diff1](https://github.com/apache/tomcat80/commit/facf983d8a7952e8d6b03204cb6610fc5f981dc6#diff-e58d7c084e48bd454de5beb71712f72d243bd0287ae2096d7b771f3a2dae6fb0R174), [diff2](https://github.com/apache/hadoop/commit/45620eee687b1093c812873a99833dab068e43d2#diff-f1ae2a4b3016a9deb983f2562d1e71372baf960cf5122d474a37a00f0307edb2R45)

Learn more: [InputStream.skip in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#skip-long-)